src/Lorentz/Contracts/BaseDAO/Types.hs

```diff
@@ -5,72 +5,208 @@
 
 module Lorentz.Contracts.BaseDAO.Types
   ( Operators
+  , mkOperators
+
   , Ledger
-  , Storage (..)
+  , LedgerValue
 
+  , Parameter (..)
+  , ProposeParams(..)
+  , Proposal (..)
+
+  , Storage (..)
+  , StorageC
   , emptyStorage
   , mkStorage
 
-  , mkOperators
   ) where
 
+import Universum
+
 import qualified Data.Map.Internal as Map
-import Text.Show (Show (..))
 
 import Lorentz
-import Lorentz.Contracts.Spec.FA2Interface
+-- import Lorentz.Contracts.Spec.FA2InterfaceCustom
+import qualified Lorentz.Contracts.Spec.FA2InterfaceCustom as FA2
 import Michelson.Runtime.GState (genesisAddress)
 
-type Operators = BigMap (Address, (Address, TokenId)) ()
+import qualified Data.Kind as Kind
 
-type Ledger = BigMap (Address, TokenId) Natural
+------------------------------------------------------------------------
+-- Operators
+------------------------------------------------------------------------
+
+type Operators = BigMap (Address, (Address, FA2.TokenId)) ()
+
+-- TODO: Needed for testing purpose
+mkOperators
+  :: Address -> Address
+  -> Address -> Address
+  -> Operators
+mkOperators owner1 operator1 owner2 operator2 =
+  BigMap $ Map.fromList
+    [ ((owner1, (operator1, 0)), ())
+    , ((owner2, (operator2, 0)), ())
+    ]
+
+------------------------------------------------------------------------
+-- Ledger
+------------------------------------------------------------------------
+
+type Ledger = BigMap LedgerKey LedgerValue
+
+type LedgerKey = (Address, FA2.TokenId)
+
+type LedgerValue = Natural
+
+------------------------------------------------------------------------
+-- Storage
+------------------------------------------------------------------------
 
 -- | Storage of the FA2 contract
-data Storage = Storage
+data Storage (proposalMetadata :: Kind.Type) = Storage
   { sLedger       :: Ledger
   , sOperators    :: Operators
   , sTokenAddress :: Address
 
   , sAdmin        :: Address
+  , sProposalIdCounter :: Natural
+
+  , sProposals :: BigMap ProposalId (Proposal proposalMetadata)
+  , sProposalsIdSortByDate :: [ProposalId] -- Needed for `flush`
+
+  -- Configuration (TODO: make it a separate record)
+  , sRejectedProposalReturnValue :: Natural
```

Comment:
Took me a while to figure out how to make the types that contain `proposalMetadata` type params work (Storage, Proposal etc ...)

This is still pretty much WIP, only `propose` entrypoint is implemented which is still incomplete.

There seems to be some changes from the spec feedback that will change most of the things here as well. 

So the MR is not yet ready for review, submitted for showing current progress only. I probably focus on adding tests for these entrypoints first instead until we got the first initial feedback for the spec.

Reply:
@gromakovsky I've updated this MR. Even though it is still WIP, I would like some of your feedback whether or not I'm on the right track, Ivan.

I have implemented:
- `propose` entrypoint: It can now check if a proposal is valid, and add to the storage. 
  - It is still in completed because it relies on `creditTo` and `debitFrom`
- `propose` tests: Current it can check for some invalid proposals.

Incompleted:
- `cProposalCheck` is current just a `Natural` which stand for the minimum amount of token required to create a proposal. I will wait for the feedback before making this a lambda.
- `vote` and other entrypoints/tests are not yet implemented.

Reply:
I just figure this out. I have update this by adding the `Config`. Let me know if this is good enough, Ivan.

Reply:
Yes, looks good now.

Reply:
Thanks, Ivan. I'll continue to implement the rest of entrypoints and adding the tests. I'll let you know when this is ready for review again.